test/test_LO.py

```diff
@@ -0,0 +1,118 @@
+import torch
+from typing import Iterator
+import functools
+import contextlib
+
+# TODO: move this into library proper
+@contextlib.contextmanager
+def no_dispatch() -> Iterator[None]:
+    guard = torch._C._DisableTorchDispatch()
+    try:
+        yield
+    finally:
+        del guard
+
+HANDLED_FUNCTIONS_RECTANGULAR = {}
+
+def implements(torch_aten_function, HANDLES_DICT):
+    """Register a torch function override for RectangularTensor.
+
+    This decorator takes a function in the torch API as a
+    parameter. Applying this decorator to a function adds that function
+    as the registered override for the torch function passed as a
+    parameter to the decorator.
+    """
+    @functools.wraps(torch_aten_function)
+    def decorator(func):
+        HANDLES_DICT[torch_aten_function] = func
+        return func
+    return decorator
+
+class RectangularTensor(torch.Tensor):
+    handled_functions = HANDLED_FUNCTIONS_RECTANGULAR
+    __slots__ = "full_rank", "well_cond"
+
+    @staticmethod
+    def __new__(cls, elem, *args, **kwargs):
+        # The wrapping tensor (RectangularTensor) is just a meta tensor, so it
+        # doesn't hold any memory (meta tensor is generally the preferred type
+        # of tensor you want to make a subclass from)...
+        r = torch.Tensor._make_subclass(cls, elem.to('meta'), elem.requires_grad)
+        # ...the real tensor is held as an element on the tensor.
+        r.elem = elem
+        r.full_rank = kwargs['full_rank'] if 'full_rank' in kwargs else True
+        r.well_cond = kwargs['well_cond'] if 'well_cond' in kwargs else True
+        r.rcond = kwargs['rcond'] if 'rcond' in kwargs else None
+        return r
+
+    def __repr__(self):
+        return f"RectangularTensor({self.elem})"
+
+    def tensor(self):
+        return self.elem
+
+    @classmethod
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        if func not in self.handled_functions:
+            return NotImplemented
+        return self.handled_functions[func](*args, **kwargs)
+
+@implements(torch.ops.aten.linalg_solve, HANDLED_FUNCTIONS_RECTANGULAR)
+def solve(A, B):
+    A_tensor = A.elem
+    # As per https://github.com/pytorch/pytorch/issues/54151
+    rcond = A.rcond if A.rcond is not None else \
+        torch.finfo(A_tensor.dtype).eps * max(A_tensor.size(-2), A_tensor.size(-1))
+    if A.full_rank:  # QR
+        return torch.linalg.lstsq(A_tensor, B, driver="gels").solution
+    else:
+        if self.well_cond:  # PQR
+            return torch.linalg.lstsq(A_tensor, B, rcond=rcond, driver="gelsy").solution
+        else:  # SVD
+            return torch.linalg.lstsq(A_tensor, B, rcond=rcond, driver="gelsd").solution
+
+HANDLED_FUNCTIONS_SQUARE = {}
+class Square(RectangularTensor):
+    handled_functions = HANDLED_FUNCTIONS_SQUARE
```

Comment:
How would we do inheritance in terms of the handled functions? For this example it doesn't matter since we're registering `torch.ops.aten.linalg_solve` for both square and rectangular tensors with different implementations, but there may be other functions that we just want to inherit. Should we allow `implements` to take a variable list of dicts and manually pass in the relevant ones? Or should we define some natural hierarchy of the handled function dicts and then find check for a registry entry in the closest parent node of that tree?

Reply:
In `LinearOperator` we deal with that simply by defining the implementations on the classes and either inherit or overwrite them, and then have the `__torch_function__` point to the implementation on the base class.

Reply:
In fact, inheritance is important even in this case, as square solvers assume that the matrix is full rank. If the matrix is not full rank, we would have to delegate to a rectangular (least squares) solver.

Reply:
Actually, this function already implements this in `solve`, where it delegates to `Rectangular` in the case above, right?